File: src/logic/AgentGrid.ts
import { Agent } from './Agent';
import { Point2 } from './core/math';

/**
 * Spatial grid for agent collision optimization
 * Uses fixed 256m cell size with Halton sequence offset to prevent persistent border issues
 */

const CELL_SIZE = 256; // meters
const WORLD_MIN_X = -10000;
const WORLD_MIN_Y = -10000;
const WORLD_MAX_X = 10000;
const WORLD_MAX_Y = 10000;
const GRID_WIDTH = Math.ceil((WORLD_MAX_X - WORLD_MIN_X) / CELL_SIZE);
const GRID_HEIGHT = Math.ceil((WORLD_MAX_Y - WORLD_MIN_Y) / CELL_SIZE);

export class AgentGrid {
    public readonly cellData: Uint16Array;
    public readonly cellOffsets: Uint32Array;
    public readonly cellCounts: Uint16Array;
    private haltonOffset: Point2 = { x: 0, y: 0 };
    private frameCounter: number = 0;
    public readonly maxAgentsPerCell: number = 256;
    public readonly totalCells: number = GRID_WIDTH * GRID_HEIGHT;

    constructor() {
        // Pre-allocate typed arrays
        this.cellData = new Uint16Array(this.totalCells * this.maxAgentsPerCell);
        this.cellOffsets = new Uint32Array(this.totalCells);
        this.cellCounts = new Uint16Array(this.totalCells);
        
        // Initialize cell offsets
        for (let i = 0; i < this.totalCells; i++) {
            this.cellOffsets[i] = i * this.maxAgentsPerCell;
        }
    }

    /**
     * Clear grid and re-index all agents with Halton offset
     */
    public clearAndReindex(agents: Agent[]): void {
        // Clear cell counts
        this.cellCounts.fill(0);

        // Generate Halton offset for this frame
        this.generateHaltonOffset();

        // Re-index all agents
        for (let i = 0; i < agents.length; i++) {
            const agent = agents[i];
            const cellIndex = this.getCellIndex(agent.coordinate);
            if (cellIndex >= 0 && cellIndex < this.totalCells) {
                const count = this.cellCounts[cellIndex];
                if (count < this.maxAgentsPerCell) {
                    const offset = this.cellOffsets[cellIndex] + count;
                    this.cellData[offset] = i;
                    this.cellCounts[cellIndex]++;
                }
            }
        }

        this.frameCounter++;
    }

    /**
     * Convert world coordinate to cell index with Halton offset
     */
    public getCellIndex(coordinate: Point2): number {
        const offsetX = coordinate.x + this.haltonOffset.x;
        const offsetY = coordinate.y + this.haltonOffset.y;
        
        const gridX = Math.floor((offsetX - WORLD_MIN_X) / CELL_SIZE);
        const gridY = Math.floor((offsetY - WORLD_MIN_Y) / CELL_SIZE);
        
        if (gridX < 0 || gridX >= GRID_WIDTH || gridY < 0 || gridY >= GRID_HEIGHT) {
            return -1;
        }
        
        return gridY * GRID_WIDTH + gridX;
    }

    /**
     * Generate Halton sequence offset for current frame
     * Uses bases 2 and 3 for 2D sequence
     */
    private generateHaltonOffset(): void {
        const halfCell = CELL_SIZE * 0.5;
        this.haltonOffset.x = this.halton(this.frameCounter, 2) * CELL_SIZE - halfCell;
        this.haltonOffset.y = this.halton(this.frameCounter, 3) * CELL_SIZE - halfCell;
    }

    /**
     * Generate Halton sequence value
     */
    private halton(index: number, base: number): number {
        let result = 0;
        let fraction = 1;
        let i = index;
        
        while (i > 0) {
            fraction /= base;
            result += (i % base) * fraction;
            i = Math.floor(i / base);
        }
        
        return result;
    }
}

File: src/wasm/agent_grid.cpp
#include "agent_grid.h"
#include "data_structures.h"
#include <cmath>
#include <vector>

const float CELL_SIZE = 256.0f;
const float WORLD_MIN_X = -10000.0f;
const float WORLD_MIN_Y = -10000.0f;
const float WORLD_MAX_X = 10000.0f;
const float WORLD_MAX_Y = 10000.0f;
const int GRID_WIDTH = static_cast<int>(ceil((WORLD_MAX_X - WORLD_MIN_X) / CELL_SIZE));
const int GRID_HEIGHT = static_cast<int>(ceil((WORLD_MAX_Y - WORLD_MIN_Y) / CELL_SIZE));
const int MAX_AGENTS_PER_CELL = 256;
const int TOTAL_CELLS = GRID_WIDTH * GRID_HEIGHT;

AgentGridData agent_grid;
Point2 halton_offset = {0.0f, 0.0f};
int frame_counter = 0;

extern AgentSoA agent_data;

// Halton sequence generator
float halton(int index, int base) {
  float result = 0.0f;
  float f = 1.0f / base;
  int i = index;
  while (i > 0) {
    result += f * (i % base);
    i = floor(i / base);
    f /= base;
  }
  return result;
}

void generate_halton_offset() {
  const float halfCell = CELL_SIZE * 0.5f;
  halton_offset.x = halton(frame_counter, 2) * CELL_SIZE - halfCell;
  halton_offset.y = halton(frame_counter, 3) * CELL_SIZE - halfCell;
}

void initialize_agent_grid(int max_agents) {
  agent_grid.cell_data.resize(TOTAL_CELLS * MAX_AGENTS_PER_CELL);
  agent_grid.cell_offsets.resize(TOTAL_CELLS);
  agent_grid.cell_counts.resize(TOTAL_CELLS);

  for (int i = 0; i < TOTAL_CELLS; i++) {
    agent_grid.cell_offsets[i] = i * MAX_AGENTS_PER_CELL;
  }
}

void clear_and_reindex_grid(int num_agents) {
  std::fill(agent_grid.cell_counts.begin(), agent_grid.cell_counts.end(), 0);

  generate_halton_offset();

  for (int i = 0; i < num_agents; i++) {
    if (!agent_data.is_alive[i]) continue;
    
    Point2 pos = agent_data.positions[i];
    int cell_index = get_cell_index(pos);

    if (cell_index >= 0 && cell_index < TOTAL_CELLS) {
      int count = agent_grid.cell_counts[cell_index];
      if (count < MAX_AGENTS_PER_CELL) {
        int offset = agent_grid.cell_offsets[cell_index] + count;
        agent_grid.cell_data[offset] = i;
        agent_grid.cell_counts[cell_index]++;
      }
    }
  }

  frame_counter++;
}

int get_cell_index(Point2 position) {
  float offset_x = position.x + halton_offset.x;
  float offset_y = position.y + halton_offset.y;

  int grid_x = static_cast<int>(floor((offset_x - WORLD_MIN_X) / CELL_SIZE));
  int grid_y = static_cast<int>(floor((offset_y - WORLD_MIN_Y) / CELL_SIZE));

  if (grid_x < 0 || grid_x >= GRID_WIDTH || grid_y < 0 || grid_y >= GRID_HEIGHT) {
    return -1;
  }

  return grid_y * GRID_WIDTH + grid_x;
} 